lib/latte.ts

```diff
@@ -0,0 +1,482 @@
+          if (currentContext && mode == 'debug') {
+            await applyCallbacks('afterEach', currentContext)
+          }
+
+          // each test gets a context that inherits from the suite-level context.
+          // This isolates each test but lets us do some setup at the suite-level for speed.
+          const previousSuiteContext =
+            currentContext && Object.getPrototypeOf(currentContext)
+          const suiteContext = await changeSuite(
+            previousSuiteContext,
+            test.suite
+          )
+
+          currentContext = Object.create(suiteContext)
+          if (!currentContext) continue
+          currentContext.currentTest = test
+          Latte.currentContext = currentContext
+
+          if (mode == 'headless') {
+            let error
+            try {
+              await applyCallbacks('beforeEach', currentContext, {
+                topDown: true,
+              })
+              await runWithTimeout(test, currentContext)
+            } catch (e) {
+              realClearTimeout(currentTimeout)
+              error = e || 'undefined error'
+            }
+
+            // AfterEach generally does cleanup. If it fails, it's unsafe to run more tests.
+            // By not catching exceptions here, we abort running and allow our chrome wrapper to reload.
+            await applyCallbacks('afterEach', currentContext)
+            Latte.onTest?.(test, error, log)
+          }
+
+          if (mode == 'debug') {
+            await applyCallbacks('beforeEach', currentContext, {
+              topDown: true,
+            })
+            await runWithTimeout(test, currentContext)
+            Latte.onTest?.(test, undefined, log)
+          }
+
+          // signal that we've finished this test, in case we're aborting
+          currentTestResolve?.()
+          whenCurrentTestFinished = currentTestResolve = null
+        }
+      } finally {
+        console.log = originalConsoleLog
+      }
+    },
+
+    // Wait for the current test, then run any `after` blocks.
+    // Cleanup only needs to be called if we're going to hot reload new code
+    cleanup: async function () {
+      if (currentContext) {
+        await applyCallbacks('afterEach', currentContext)
+        await changeSuite(currentContext)
+        currentContext = null
+      }
+    }
+  }
+  window.Latte = Latte
+
+  // create the root describe block
+  let current = window.Latte.root
+
+  function describe(name: string, fn: () => void): void {
+    const previous = current
+    current = {
+      parent: previous,
+      name: name,
+      children: [],
+      tests: [],
+      depth: previous.depth + 1,
+      testCount: 0,
+      fullName: [previous.fullName, name].filter((s) => !!s).join(' '),
+      before: [],
+      beforeEach: [],
+      after: [],
+      afterEach: [],
+      helpers: {},
+      // eslint-disable-next-line @typescript-eslint/no-empty-function
+      timeout: () => {}, // TODO
```

Comment:
Acknowledging that this was here before the TS conversion - what's this TODO about?

Reply:
I believe this relates to the timeout fn we have defined in some of the tests itself. Idk the purpose, my best guess was maybe something like the extendRemoteTimeout I implemented a couple weeks ago 🤷